zebra-state/src/service/non_finalized_state.rs

```diff
@@ -87,29 +94,54 @@ impl NonFinalizedState {
             );
         }
 
-        let mut parent_chain = self
-            .take_chain_if(|chain| chain.non_finalized_tip_hash() == parent_hash)
-            .or_else(|| {
-                self.chain_set
-                    .iter()
-                    .find_map(|chain| chain.fork(parent_hash))
-                    .map(Box::new)
-            })
-            .expect("commit_block is only called with blocks that are ready to be commited");
-
-        parent_chain.push(prepared);
+        let mut parent_chain =
+            match self.take_chain_if(|chain| chain.non_finalized_tip_hash() == parent_hash) {
+                Some(chain) => chain,
+                None => Box::new(
+                    self.chain_set
+                        .iter()
+                        .find_map(|chain| {
+                            chain
+                                .fork(parent_hash, finalized_tip_history_tree)
+                                .transpose()
+                        })
+                        .expect(
+                            "commit_block is only called with blocks that are ready to be commited",
+                        )?,
+                ),
+            };
```

Comment:
Readability: Can you split `parent_chain` into its own method?

That would keep `commit_block` short, and make they validation and chain operation order clearer.